### PR TITLE
Knowledge distillation

### PR DESCRIPTION
**Pull request by SKholkin**
Knowledge distillation future implementation. Feel free to leave your questions and comments.

### CODE REVIEW

File: nncf/compression_method_api.py [hunk 1]
```diff
@@ -52,21 +52,21 @@ class PTCompressionLoss(nn.Module, CompressionLoss):
     and fully-connected layers to construct the loss function.
     """
 
-    def calculate(self) -> torch.Tensor:
+    def calculate(self, input_=None, target=None) -> torch.Tensor:
```
Comment (SKholkin) [inline nncf/compression_method_api.py:55]: @alexsu52 
@vshampor 
KDLossCalculator requires API changes because we need to get arguments (input_  and target) inside KDLossCalculator module to calculate loss. So should this changes need to be done for  PyTorch Compression API only or for Unified API too? In other words: will this changes affect TF Loss calculation policies?

Comment (vshampor) [reply nncf/compression_method_api.py:55]: I suppose that since knowledge distillation loss fundamentally requires some sort of inputs and outputs, then the change must be done on the unified API level.

Comment (SKholkin) [reply nncf/compression_method_api.py:55]: + @daniil-lyakhov

Comment (dupeljan) [reply nncf/compression_method_api.py:55]: NNCF TF compression loss have completely different api and do all work inside `call` function.  I think this doesn't affect TF and will simply unify in future. CC: @alexsu52

Comment (daniil-lyakhov) [reply nncf/compression_method_api.py:55]: NNCF TF compression loss have completely different api and do all work inside `call` function.  I think this doesn't affect TF and will simply unify in future. CC: @alexsu52

**Comment by vanyalzr** [comment]
@SKholkin please do introduce an option to switch from MSE KD loss to KL divergence loss and vice versa. Also we should think about how to handle models with arbitrary outputs, for example in our OD sample we have SSD models returning anchor box coordinates which should not be a part of KD loss, and we are not able to infer it from e.g. the `requires_grad` attribute.

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by FyodorK** [comment]
Jenkins please stop all builds

**Comment by FyodorK** [comment]
Jenkins please retry a build

File: nncf/torch/kd_loss.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        """
+
+        compressed_model_outputs_struct = []
+        orig_model_outputs_struct = []
+        OperatorInput.nested_object_paths_generator([compressed_model_outputs], compressed_model_outputs_struct)
+        OperatorInput.nested_object_paths_generator([orig_model_outputs], orig_model_outputs_struct)
+        compressed_model_loss_nested_obj_paths = list(filter(lambda x: self._is_loss(x.getter()),
+                                                             compressed_model_outputs_struct))
+        compressed_model_loss_outputs = list(map(lambda x: x.getter(), compressed_model_loss_nested_obj_paths))
+        orig_model_loss_outputs = list(map(lambda x: x.getter(), filter(
```
Comment (ljaljushkin) [inline nncf/torch/kd_loss.py:63]: please break this and similar complex constructions into small parts for readability

File: nncf/torch/kd_loss.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            partial(self.match_func, to_match_with=compressed_model_loss_nested_obj_paths), orig_model_outputs_struct)))
```
Comment (ljaljushkin) [inline nncf/torch/kd_loss.py:64]: Don't think you need to make it so general with partial and static function.
Why not just defining specific lambda function with `compressed_model_loss_nested_obj_paths` instead of `to_match_with`? It would be more readable.

File: nncf/torch/dynamic_graph/utils.py [hunk 4]
```diff
@@ -13,6 +13,7 @@
 
 from graphviz import Digraph
 
+
```
Comment (ljaljushkin) [inline nncf/torch/dynamic_graph/utils.py:16]: please remove this change

File: nncf/config/schema.py [hunk 5]
```diff
@@ -713,19 +713,36 @@ def with_attributes(schema: Dict, **kwargs) -> Dict:
...
 
+KNOWLEDGE_DISTILLATION_ALGO_NAME_IN_CONFIG = 'knowledge_distillation'
+KNOWLEDGE_DISTILLATION_SCHEMA = {
+    **BASIC_COMPRESSION_ALGO_SCHEMA,
+    "properties": {
+        "algorithm": {
+            "const": KNOWLEDGE_DISTILLATION_ALGO_NAME_IN_CONFIG
+        },
+        "scale": with_attributes(_NUMBER,
+                                 description="Scale parameter for kdloss"),
```
Comment (ljaljushkin) [inline nncf/config/schema.py:724]: not clear why this scale is needed. please describe it in more details

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

**Comment by SKholkin** [comment]
@vshampor please leave some comments

**Comment by FyodorK** [comment]
Jenkins please retry a build

File: examples/torch/object_detection/layers/modules/ssd_head.py [hunk 6]
```diff
@@ -90,6 +90,7 @@ def forward(self, features, image_tensor):
 
         with no_nncf_trace():
             priors = self.prior_box(features, image_tensor).to(loc.device)
+        priors = priors.detach()
```
[the remark this answers is not in the document]
Comment (SKholkin) [reply examples/torch/object_detection/layers/modules/ssd_head.py:93]: Without detach priors would be model outputs with requires_grad=True, but actually priors are not used in loss calculations directly so they are not included in backpropagation graph (that is coded inside MultiBoxLoss). So if model outputs shouldn't be used for backpropagation they should have requires_grad=False. 
`priors = priors.detach()` excludes priors from a backpropagation graph.
Also because KD logic requires differentiation of model outputs with requires_grad=True than we don't want model outputs that shouldn't be differentiated have requires_grad=True (that would cause an error).

File: nncf/config/schema.py (hunk 5 above)
Comment (SKholkin) [reply nncf/config/schema.py:724]: Scale is just a FP variable to multiply KDLoss value to.


`return self.scale * reduce(            lambda kd_loss, loss_tensors: kd_loss + self.kdloss_fn(loss_tensors[0], loss_tensors[1]),            zip(orig_model_loss_outputs, compressed_model_loss_outputs), 0)`


In other words:
KDLoss = scale * KDLoss

Comment (SKholkin) [reply nncf/config/schema.py:724]: Actually this option is not necessary or frequently used and could be abandoned.

File: examples/torch/object_detection/layers/modules/ssd_head.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply examples/torch/object_detection/layers/modules/ssd_head.py:93]: Yes we discussed it. for me it's clear, but not for other developers and users. I meant adding comment in the code 😄

File: nncf/config/schema.py (hunk 5 above)
Comment (ljaljushkin) [reply nncf/config/schema.py:724]: Also meant comment for the user in the code.
But if it's not useful we can remove it

Comment (vanyalzr) [reply nncf/config/schema.py:724]: I'd leave it but rename to something self-explanatory like `kd_loss_weight_in_total_loss`

File: nncf/torch/kd_loss.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+class KnowledgeDistillationController(PTCompressionAlgorithmController):
+    def compression_level(self) -> CompressionLevel:
+        return CompressionLevel.FULL
+
+    def __init__(self, target_model, original_model, scale, type):
+        super().__init__(target_model)
+        original_model.train()
+        self._scheduler = BaseCompressionScheduler()
+        self._loss = KDLossCalculator(target_model=target_model, original_model=original_model, scale=scale, type=type)
+        target_model.enable_knowledge_distillation(original_model, self._loss.calculate_kd_loss)
```
[the remark this answers is not in the document]
Comment (SKholkin) [reply nncf/torch/kd_loss.py]: `KDLossHandler `class for KD logic encapsulation has been added pleas have a look

File: examples/torch/object_detection/layers/modules/ssd_head.py [hunk 8]
```diff
@@ -90,6 +90,8 @@ def forward(self, features, image_tensor):
 
         with no_nncf_trace():
             priors = self.prior_box(features, image_tensor).to(loc.device)
+        # non differential model outputs should be marked by requires_grad=False
```
Comment (ljaljushkin) [inline examples/torch/object_detection/layers/modules/ssd_head.py:93]: I'd rather you mention connection with KD.

File: nncf/config/schema.py [hunk 9]
```diff
@@ -713,19 +713,34 @@ def with_attributes(schema: Dict, **kwargs) -> Dict:
...
 }
 
+KNOWLEDGE_DISTILLATION_ALGO_NAME_IN_CONFIG = 'knowledge_distillation'
+KNOWLEDGE_DISTILLATION_SCHEMA = {
+    **BASIC_COMPRESSION_ALGO_SCHEMA,
+    "properties": {
+        "algorithm": {
+            "const": KNOWLEDGE_DISTILLATION_ALGO_NAME_IN_CONFIG
+        },
+        "type": with_attributes(_STRING, description="Type of KDLoss (mse/softmax)")
```
Comment (ljaljushkin) [inline nncf/config/schema.py:723]: please avoid contractions in the documentation. Names like `KDLoss` is ok for the code, but not for the users reading documentation. Please go through all added docstrings and fix it.

File: nncf/torch/kd_loss.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+                return mse(ref_outputs, compressed_model_outputs)
+        else:
+            raise ValueError('Choose between mse/softmax options for Knowledge Distillation')
+        self._kd_loss_handler = target_model.create_kd_loss_handler(original_model, partial(
+            KnowledgeDistillationLoss.calculate_kd_loss,
+            device=next(target_model.parameters()).device,
+            kd_loss_fn=kd_loss_fn))
+
+    @staticmethod
+    def calculate_kd_loss(compressed_model_outputs, orig_model_outputs, device, kd_loss_fn):
```
Comment (ljaljushkin) [inline nncf/torch/kd_loss.py:42]: IMHO, `kd_loss` in the name is redundant

File: nncf/torch/kd_loss.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        """
+        Calculates KD Loss value from compressed_model_outputs and orig_model_outputs. First uses
```
Comment (ljaljushkin) [inline nncf/torch/kd_loss.py:44]: the same comment about contractions in the documentation

File: nncf/torch/kd_loss.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        nested_object_paths_generator to unpack input containers and numerate contents inside them.
+        Than checks compressed_model_outputs unpacked container for loss tensors (requires_grad=True)
+        and maps extracted structure of loss tensors to orig_model_outputs.
+        Finally computes KD loss with extracted loss tensors.
+
+        :param compressed_model_outputs: Output tensors of compressed model can be any type of container with
```
Comment (ljaljushkin) [inline nncf/torch/kd_loss.py:50]:
```suggestion
        :param compressed_model_outputs: Output tensors of compressed model can be any type of container that iterates over internal items in the deterministic way.
```

File: tests/torch/test_kd.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+    if isinstance(config['compression'], dict):
+        config['compression'] = [config['compression']]
+    config['compression'].append({
+        'algorithm': 'knowledge_distillation',
+        'type': 'mse'
+    })
+    return config
+
+
+def get_sparsity_config_with_sparsity_init(config, sparsity_init=0.5):
```
Comment (ljaljushkin) [inline tests/torch/test_kd.py:31]: can this config be reused from sparsity tests?

File: tests/torch/test_kd.py [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+        for tensor1, tensor2 in zip(outputs, kd_outputs):
+            if tensor1.requires_grad:
+                loss_outputs.append((tensor1, tensor2))
+
+        reference_kd_loss = sum([mse(item[0], item[1]) for item in loss_outputs])
+        actual_kd_loss = compression_ctrl.loss()
+        assert torch.allclose(reference_kd_loss, actual_kd_loss)
+
+
+def test_kd_outputs_contrainers_parsing():
```
Comment (ljaljushkin) [inline tests/torch/test_kd.py:189]: containers?

File: tests/torch/quantization/test_quantization_helpers.py [hunk 15]
```diff
@@ -33,17 +33,18 @@ def compare_multi_gpu_dump(config, dump_dir, get_path_by_rank_fn):
...
+        self._len = num_samples
         super().__init__()
 
     def __getitem__(self, index):
         dummy_input = torch.ones(self.input_size) * (self.rank - 1) * 3
         return dummy_input, torch.ones(1)
 
     def __len__(self):
-        return 100
+        return self._len
```
Comment (ljaljushkin) [inline tests/torch/quantization/test_quantization_helpers.py:47]: 👍

File: nncf/torch/nncf_network.py [hunk 16]
```diff
@@ -469,6 +507,9 @@ def strip_fn(tensor: TracedTensor) -> torch.Tensor:
         kwargs = objwalk(kwargs, is_traced_tensor_predicate, strip_fn)
         return args, kwargs
 
+    def create_kd_loss_handler(self, kd_original_model, calculate_fn):
```
Comment (ljaljushkin) [inline nncf/torch/nncf_network.py:510]: please add docstring according to the CodeStyle

File: nncf/torch/nncf_network.py [hunk 17]
```diff
@@ -426,6 +459,8 @@ def __init__(self, module, input_infos: List[ModelInputInfo],
 
         self._compressed_context = TracingContext()
 
+        self._compressed_context.register_global_buffer(STORAGE_DEVICE, device)
```
Comment (ljaljushkin) [inline nncf/torch/nncf_network.py:462]: why not to move it to __init__ of `KDLossHandler` by passing required device from NNCFNetwork?

File: nncf/torch/kd_loss.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+
+    def build_controller(self, target_model):
+        return KnowledgeDistillationController(target_model, self.original_model, self.type_)
+
+    def initialize(self, model: NNCFNetwork) -> None:
+        pass
+
+
+class KnowledgeDistillationController(PTCompressionAlgorithmController):
+    def compression_level(self) -> CompressionLevel:
```
Comment (ljaljushkin) [inline nncf/torch/kd_loss.py:126]: please declare public methods after __init__

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: nncf/torch/dynamic_graph/op_input_processing.py [hunk 19]
```diff
@@ -52,17 +52,17 @@ def __init__(self, op_args, op_kwargs):
         self._index = OrderedDict()  # type: Dict[int, InputIndexEntry]
 
         op_args_index_entries = []
-        self._nested_object_paths_generator(self.op_args, op_args_index_entries,
```
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply nncf/torch/dynamic_graph/op_input_processing.py:55]: move this function to `nested_object_traversal.py` with `obj_walk` and `maybe_get_iterator` @vshampor

**Comment by SKholkin** [comment]
@vshampor Please leave some comments

File: examples/torch/object_detection/layers/modules/ssd_head.py [hunk 20]
```diff
@@ -90,6 +90,9 @@ def forward(self, features, image_tensor):
 
         with no_nncf_trace():
             priors = self.prior_box(features, image_tensor).to(loc.device)
+        # Through usage of Knowledge Distillation Algo priors were incorrectly differentiated what was causing on error.
+        # Non differential model outputs should be marked by requires_grad=False.
```
Comment (ljaljushkin) [inline examples/torch/object_detection/layers/modules/ssd_head.py:94]: Sorry for being picky, but I don't like this message.
Maybe in the next PR, but please try to explain the assumptions of KD about model outputs in more details. 
After reading this message still not clear the root cause of issue. 
Also it makes sense explain why you are doing detach instead of requires_grad=True

please create task for it, if you are not going to close it in this PR

@vanyalzr what do you think? does it make sense from your point of view?

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: nncf/torch/dynamic_graph/context.py [hunk 21]
```diff
@@ -50,6 +53,35 @@ def __hash__(self):
         return hash(str(self))
 
 
+class OperatorInput:
```
[the remark this answers is not in the document]
Comment (vshampor) [reply nncf/torch/dynamic_graph/context.py:56]: Actually I think I moved it out from `context.py` into `op_input_processing.py` before, so surely it shouldn't return to `context.py`.

File: nncf/compression_method_api.py (hunk 1 above)
Comment (ljaljushkin) [reply nncf/compression_method_api.py]: created 58573 task to explore TF Part

File: examples/torch/object_detection/layers/modules/ssd_head.py (hunk 20 above)
Comment (SKholkin) [reply examples/torch/object_detection/layers/modules/ssd_head.py:94]: Maybe we could explain the assumptions of KD about model outputs in more general way than in `ssd_head.py`. So it would be a part of more general docs writing task. For now just leave like this or maybe even delete KD reference as explanation of `priors = priors.detach()`

File: nncf/torch/dynamic_graph/context.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (SKholkin) [reply nncf/torch/dynamic_graph/context.py]: Moved `OperatorInput` back to `op_input_processing.py`

**Comment by SKholkin** [comment]
Jenkins please retry a build

**Comment by SKholkin** [comment]
Jenkins please retry a build

**Comment by SKholkin** [comment]
Jenkins please retry a build

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by FyodorK** [comment]
Jenkins please stop this build

**Comment by FyodorK** [comment]
Jenkins please stop all builds

**Comment by SKholkin** [comment]
Jenkins please retry a build

**Comment by SKholkin** [comment]
Jenkins please retry a build

**Comment by SKholkin** [comment]
Jenkins please stop all builds

**Comment by SKholkin** [comment]
Jenkins please retry a build

File: nncf/torch/nested_objects_traversal.py [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+
+    def __call__(self, value):
+        tmp_list = list(self._current_tuple)
+        tmp_list[self._idx_to_set] = value
+        new_tuple = tuple(tmp_list)
+        self._current_tuple = new_tuple
+        self._previous_level_setter(new_tuple)
+
+
+def nested_object_paths_generator(obj, out_entries_list, path=(), memo=None, previous_level_setter=None):
```
Comment (vshampor) [inline nncf/torch/nested_objects_traversal.py:60]: This should not be a free function, and a function with side effects. If you need to reuse the path generation functionality  in more places than one, then encapsulate it into a proper class object with well-defined inputs and outputs and no side effects. Use the object wherever needed.

You can call the new class `NestedObjectIndex`, storing and accessing `IndexEntry` objects like it was done before for inputs, and reuse it, correspondingly, wherever the flat nested object indexing is required, such as a) for input wrapping and input port detection, b) for differential output deduction, and c) (in the future) for output port detection.

File: nncf/torch/nncf_network.py [hunk 23]
```diff
@@ -323,6 +323,42 @@ def __hash__(self):
...
 
 
+class KnowledgeDistillationLossHandler(nn.Module):
+    """
+    Encapsulates knowledge distillation logic. Controls loss calculation, its proper storing in case of parallel
+        execution on multiple devices and provides API for external knowledge distillation algorithm.
+    """
+    KD_LOSS_STORAGE_NAME = 'kd_loss'
+    KD_STORAGE_DEVICE = 'kd_storage_device'
+    def __init__(self, context, kd_original_model, calculate_kd_loss_fn, storage_device):
```
Comment (vshampor) [inline nncf/torch/nncf_network.py:333]: Type hints

File: nncf/torch/nncf_network.py [hunk 24]
```diff
@@ -323,6 +323,42 @@ def __hash__(self):
...
+
+    def get_kdloss(self):
+        if self.is_enabled is not None:
+            return self._compressed_context.global_buffer_store[self.KD_LOSS_STORAGE_NAME]
+        return 0
+
+    def forward(self, inputs, *args, **kwargs):
+        kd_model = self._kd_original_model
+        with torch.no_grad():
+            kd_outputs = kd_model(*args, **kwargs)
```
Comment (vshampor) [inline nncf/torch/nncf_network.py:354]: Rename to `kd_original_model`

File: nncf/torch/nncf_network.py [hunk 25]
```diff
@@ -323,6 +323,42 @@ def __hash__(self):
...
+        self._compressed_context.register_global_buffer(self.KD_STORAGE_DEVICE, storage_device)
+
+    def zero_kdloss(self):
+        if self.is_enabled is not None:
+            self._compressed_context.global_buffer_store[self.KD_LOSS_STORAGE_NAME] = []
+
+    def get_kdloss(self):
+        if self.is_enabled is not None:
+            return self._compressed_context.global_buffer_store[self.KD_LOSS_STORAGE_NAME]
+        return 0
```
Comment (vshampor) [inline nncf/torch/nncf_network.py:349]: Type hints. Make sure that the PyTorch int `0` is what you actually want to be returned - shouldn't this be `torch.Tensor` instead?

File: nncf/torch/nncf_network.py [hunk 26]
```diff
@@ -440,6 +480,22 @@ def strip_fn(tensor: TracedTensor) -> torch.Tensor:
         kwargs = objwalk(kwargs, is_traced_tensor_predicate, strip_fn)
         return args, kwargs
 
+    def create_knowledge_distillation_loss_handler(self, kd_original_model, calculate_fn):
```
Comment (vshampor) [inline nncf/torch/nncf_network.py:483]: Type hints

File: tests/common/graph/test_graph_matching.py [hunk 27]
```diff
@@ -75,4 +75,5 @@ def test_two_matches():
     ref_graph.add_edge('5', '6')
 
     matches = find_subgraphs_matching_pattern(ref_graph, pattern)
-    assert matches in [[['1', '2'], ['5', '6']]]
+    matches.sort()
+    assert matches == [['1', '2'], ['5', '6']]
```
Comment (vshampor) [inline tests/common/graph/test_graph_matching.py:79]: Rebase to remove this

File: tests/torch/test_algo_common.py [hunk 28]
```diff
@@ -490,6 +491,8 @@ def test_target_device_is_propagated_to_algos(mocker, algo_name, target_device):
         },
         "target_device": target_device
     })
+    if COMPRESSION_ALGORITHMS.get(algo_name) == KnowledgeDistillationBuilder:
```
Comment (vshampor) [inline tests/torch/test_algo_common.py:494]: Just do the comparison with the name of the algo as it is present in the config, no need to involve class objects.

File: tests/torch/test_knowledge_distillation.py [hunk 29]
```diff
@@ -0,0 +1,234 @@
...
+                                                  batch_size=torch.cuda.device_count())
+    compression_ctrl.scheduler.epoch_step()
+    for _, (input_, __) in enumerate(mock_dataloader):
+        input_ = input_.to(next(model.parameters()).device)
+        outputs = model(input_)
+        kd_outputs = dumped_orig_model(input_)
+
+        reference_kd_loss = kd_loss_fn(kd_outputs, outputs)
+        actual_kd_loss = compression_ctrl.loss()
+        assert torch.allclose(reference_kd_loss, actual_kd_loss)
```
Comment (vshampor) [inline tests/torch/test_knowledge_distillation.py:234]: Set up type hints across this file as well wherever it makes sense to do so.

File: tests/torch/test_models/synthetic.py [hunk 30]
```diff
@@ -148,6 +148,39 @@ def forward(self, x):
         return xa, xb, xc, xd
 
 
+class PartlyNonDifferentialOutputsModel(nn.Module):
+    def __init__(self, input_size=None):
+        super().__init__()
+        self.input_size = [1, 1, 4, 4] if input_size is None else input_size
+        self.Conv1 = torch.nn.Conv2d(in_channels=self.input_size[1], out_channels=1, kernel_size=3)
+        self.Conv2_1 = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=3)
+        self.Conv2_2 = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=3)
```
Comment (vshampor) [inline tests/torch/test_models/synthetic.py:157]: Variables in Python should be lower-cased, you can see that something is wrong already in the Github's code highlighter.

File: nncf/torch/knowledge_distillation_loss.py [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+from nncf.torch.compression_method_api import PTCompressionAlgorithmBuilder
+from nncf.torch.compression_method_api import PTCompressionAlgorithmController
+from nncf.api.compression import CompressionLoss, CompressionScheduler, CompressionStage
+from nncf.torch.algo_selector import COMPRESSION_ALGORITHMS
+from nncf.torch.compression_method_api import PTCompressionLoss
+from nncf.torch.nested_objects_traversal import nested_object_paths_generator
+
+
+class KnowledgeDistillationLoss(PTCompressionLoss):
+    def __init__(self, target_model: NNCFNetwork, original_model, kd_type):
```
Comment (vshampor) [inline nncf/torch/knowledge_distillation_loss.py:21]: Type hints everywhere

File: nncf/torch/knowledge_distillation_loss.py [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+        Calculates knowledge distillation loss value from compressed_model_outputs and orig_model_outputs. First uses
+        nested_object_paths_generator to unpack input containers and numerate contents inside them.
+        Than checks compressed_model_outputs unpacked container for loss tensors (requires_grad=True)
+        and maps extracted structure of loss tensors to orig_model_outputs.
+        Finally computes knowledge distillation loss with extracted loss tensors.
+
+        :param compressed_model_outputs: Output tensors of compressed model can be any type of container that
+            iterates over internal items in the deterministic way.
+        :param orig_model_outputs: Output tensors of original model (used for distillation) can be any type of
+            container with deterministic detour.
```
Comment (vshampor) [inline nncf/torch/knowledge_distillation_loss.py:52]: What did you mean by "detour" here? Reword this to be consistent with the docstring for the first param.

File: nncf/torch/knowledge_distillation_loss.py [hunk 33]
```diff
@@ -0,0 +1,151 @@
...
+from nncf.torch.nncf_network import NNCFNetwork
+from nncf.torch.compression_method_api import PTCompressionAlgorithmBuilder
+from nncf.torch.compression_method_api import PTCompressionAlgorithmController
+from nncf.api.compression import CompressionLoss, CompressionScheduler, CompressionStage
+from nncf.torch.algo_selector import COMPRESSION_ALGORITHMS
+from nncf.torch.compression_method_api import PTCompressionLoss
+from nncf.torch.nested_objects_traversal import nested_object_paths_generator
+
+
+class KnowledgeDistillationLoss(PTCompressionLoss):
```
Comment (vshampor) [inline nncf/torch/knowledge_distillation_loss.py:20]: Should state explicitly somewhere in a docstring that only the inputs of the model that were consumed by the model during the latest `forward` call will be used to determine original model output to be compared with the compressed model's current output. This is a limitation of the algorithm, so we need to make sure that the user understands it.

File: nncf/torch/nncf_network.py [hunk 34]
```diff
@@ -440,6 +480,22 @@ def strip_fn(tensor: TracedTensor) -> torch.Tensor:
...
+        """
+        Creates KnowledgeDistillationLossHandler instance for enabling Knowledge Distillation feature.
+            Also returns created KnowledgeDistillationLossHandler for control over Knowledge Distillation logic.
+
```
Comment (vshampor) [inline nncf/torch/nncf_network.py:487]: Add same note about the KD loss handler only representing the KD loss corresponding to the inputs of the latest `NNCFNetwork`'s forward call.

File: nncf/torch/nncf_network.py [hunk 35]
```diff
@@ -323,6 +323,42 @@ def __hash__(self):
         return hash(str(self))
 
 
+class KnowledgeDistillationLossHandler(nn.Module):
```
Comment (vshampor) [inline nncf/torch/nncf_network.py:326]: Can this be separated into an own .py file?

**Review by vshampor — CHANGES_REQUESTED** [review]

**Comment by SKholkin** [comment]
Jenkins please stop all builds

**Comment by SKholkin** [comment]
Jenkins please retry a build

File: tests/torch/test_algo_common.py (hunk 28 above)
Comment (SKholkin) [reply tests/torch/test_algo_common.py:494]: The thing is I can't get the name of KnowledgeDistillation algo from `KnowledgeDistillationBuilder` (except for `KnowledgeDistillationBuilder._registered_name`) so comparation between class objects seems appropriate.
`if algo_name == NoCompressionAlgorithmBuilder.__name__` works because `NoCompressionAlgorithmBuilder.__name__ == NoCompressionAlgorithmBuilder._registered_name` which is not the case for `KnowledgeDistillationBuilder`.

Or is it ok to `if algo_name == 'knowledge_distillation'`?

Comment (vshampor) [reply tests/torch/test_algo_common.py:494]: Indeed, should be enough to do `if algo_name == 'knowledge_distillation'`

Comment (ljaljushkin) [reply tests/torch/test_algo_common.py:494]: FYI, after merging my PR `name` property will be available: 
https://github.com/openvinotoolkit/nncf/pull/682/files#diff-e8f054f8808cc9acc32c920dcec746b3f541b8448922f853419fdbe1d2462debR156-R158

we shouldn't forget to adapt it to the newer code base.

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by FyodorK** [comment]
Jenkins please stop all builds

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by FyodorK** [comment]
Jenkins please retry a build

**Comment by SKholkin** [comment]
Jenkins please retry a build

**Comment by SKholkin** [comment]
Jenkins please stop all builds

**Comment by SKholkin** [comment]
Jenkins please retry a build

**Comment by SKholkin** [comment]
Jenkins please retry a build

**Comment by SKholkin** [comment]
Jenkins please stop all builds

File: nncf/torch/nested_objects_traversal.py (hunk 22 above)
Comment (vshampor) [reply nncf/torch/nested_objects_traversal.py]: Great job.

**Review by vshampor — APPROVED** [review]

File: nncf/torch/knowledge_distillation/algo.py [hunk 36]
```diff
@@ -0,0 +1,64 @@
+from copy import deepcopy
```
Comment (alexsu52) [inline nncf/torch/knowledge_distillation/algo.py:1]: Please, add a license here and another files.

Comment (ljaljushkin) [reply nncf/torch/knowledge_distillation/algo.py:1]: Let's resolve it in a separate PR. created task for it #819 
Sergei can do it in the PR with docs and e2e test.

**Review by ljaljushkin — APPROVED** [review]
I'd merge it. licences, documentation and e2e test can be done separately.

**Review by alexsu52 — APPROVED** [review]
